lib/Model/Pass/PurgeUnnamedAndUnreachableTypes.cpp

```diff
@@ -76,14 +76,26 @@ void model::purgeTypesImpl(TupleTree<model::Binary> &Model) {
   // Remember those types we want to preserve.
   if constexpr (PruneAllUnusedTypes) {
     for (const auto &Function : Model->Functions()) {
-      if (not Function.Prototype().empty()) {
+      if (not Function.Prototype().empty())
         ToKeep.insert(const_cast<Type *>(Function.Prototype().get()));
-      }
+      if (not Function.StackFrameType().empty())
+        ToKeep.insert(const_cast<Type *>(Function.StackFrameType().get()));
     }
 
+    for (const auto &DynamicFunction : Model->ImportedDynamicFunctions())
+      if (not DynamicFunction.Prototype().empty())
+        ToKeep.insert(const_cast<Type *>(DynamicFunction.Prototype().get()));
+
+    for (const auto &Segment : Model->Segments())
+      if (not Segment.Type().empty())
+        ToKeep.insert(const_cast<Type *>(Segment.Type().get()));
+
     for (UpcastablePointer<model::Type> &T : Model->Types()) {
       TypeToNode[T.get()] = TypeGraph.addNode(NodeData{ T.get() });
     }
+
+    if (not Model->DefaultPrototype().empty())
+      ToKeep.insert(const_cast<Type *>(Model->DefaultPrototype().get()));
```

Comment:
I think the whole semantics of the `PruneAllUnusedTypes == true` code paths is dubious and should be dropped or reviewed.

In any case, this change is not probably what we want. It tries to mirror (by hand) what we already do (automatically) below in the `!PruneAllUnusedTypes` code path (the `visitTupleExcept` business).

If you want this branch to go in, leave this commit out.

Reply:
Eh without this commit the tests don't pass, because types are purged too aggressively and we're left with a model that doesn't verify with the new stricter version of `verify` that this patchset introduces.

I realize the proposed change is not correct, but at the moment I'm very overcommitted with other more pressing stuff. Could you please try and remove the commit and take a look yourself when you have time?

Otherwise I'm gonna have to leave this sit there for a while before I can look at it.

Thanks

Reply:
I see that you've pushed a commit to fix this. I'm closing the conversation.